test/matchers/http.go

```
@@ -42,25 +44,61 @@ func HavePartialResponseBody(substring string) types.GomegaMatcher {
 	})
 }
 
+// HaveOkResponseWithHeaders expects an 200 response with a set of headers that match the provided headers
+func HaveOkResponseWithHeaders(headers map[string]interface{}) types.GomegaMatcher {
+	return HaveHttpResponse(&HttpResponse{
+		StatusCode: http.StatusOK,
+		Body:       gomega.BeEmpty(),
+		Headers:    headers,
+	})
+}
+
 // HttpResponse defines the set of properties that we can validate from an http.Response
 type HttpResponse struct {
 	// StatusCode is the expected status code for an http.Response
+	// Required
 	StatusCode int
 	// Body is the expected response body for an http.Response
 	// Body can be of type: {string, bytes, GomegaMatcher}
+	// Optional: If not provided, defaults to an empty string
 	Body interface{}
+	// Headers is the set of expected header values for an http.Response
+	// Each header can be of type: {string, GomegaMatcher}
+	// Optional: If not provided, does not perform header validation
+	Headers map[string]interface{}
+	// Custom is a generic matcher that can be applied to validate any other properties of an http.Response
+	// Optional: If not provided, does not perform additional validation
+	Custom types.GomegaMatcher
 }
 
 // HaveHttpResponse returns a GomegaMatcher which validates that an http.Response contains
 // particular expected properties (status, body..etc)
 // If an expected body isn't defined, we default to expecting an empty response
+// NOTE TO DEVELOPERS: The underlying matchers cache properties, so if you are wrapping this
+// matcher in an Eventually, be sure to create a new matcher each iteration.
```

Since it seems like reuse is an implementation error but we own the sub match call do we think we should store an extra attr on MatchHttpResponseMatcher and use it to work around the error?

---

I like the idea! Though I'm not sure I entirely follow. Are you proposing modifying the match implementation to construct a new body matcher each time that Match is called?

---

Updated to include an internal attribute that we use to fail loudly if a match has been evaluated more than once. I moved this comment to that failure line. Let me know if this is clear or if there's more we can do to help developers understand